Using ROW_NUMBER OVER PARTITION BY with WHERE clause throws an UnsupportedFeatureException

**CrateDB version**: 4.5.1

**Problem description**:

When adding a WHERE clause on the row number for `ROW_NUMBER OVER PARTITION BY` you get a `UnsupportedFeatureException[Can't handle Symbol [WindowFunction: row_number() OVER (PARTITION BY INPUT(0))]]`

**Steps to reproduce**:

Example query:
```
SELECT * FROM (
  SELECT col1, ROW_NUMBER() OVER(PARTITION by col1) as row_num
 FROM (VALUES('x'), ('y'), ('z')) t1) t2 
WHERE row_num = 2 LIMIT 100;
```

<details>
  <summary>
stack trace (click to expand)

```
io.crate.exceptions.UnsupportedFeatureException: Can't handle Symbol [WindowFunction: row_number() OVER (PARTITION BY INPUT(0))]
```
</summary>

```
	at io.crate.exceptions.SQLExceptions.esToCrateException(SQLExceptions.java:139)
	at io.crate.exceptions.SQLExceptions.prepareForClientTransmission(SQLExceptions.java:126)
	at io.crate.rest.action.SqlHttpHandler.sendResponse(SqlHttpHandler.java:162)
	at io.crate.rest.action.SqlHttpHandler.lambda$channelRead0$0(SqlHttpHandler.java:118)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883)
	at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2321)
	at io.crate.rest.action.SqlHttpHandler.channelRead0(SqlHttpHandler.java:116)
	at io.crate.rest.action.SqlHttpHandler.channelRead0(SqlHttpHandler.java:79)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.crate.protocols.http.HttpBlobHandler.channelRead0(HttpBlobHandler.java:166)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.crate.auth.HttpAuthUpstreamHandler.handleHttpChunk(HttpAuthUpstreamHandler.java:136)
	at io.crate.auth.HttpAuthUpstreamHandler.channelRead0(HttpAuthUpstreamHandler.java:85)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Thread.java:831)
Caused by: java.lang.UnsupportedOperationException: Can't handle Symbol [WindowFunction: row_number() OVER (PARTITION BY INPUT(0))]
	at io.crate.expression.BaseImplementationSymbolVisitor.visitSymbol(BaseImplementationSymbolVisitor.java:101)
	at io.crate.expression.BaseImplementationSymbolVisitor.visitSymbol(BaseImplementationSymbolVisitor.java:43)
	at io.crate.expression.symbol.SymbolVisitor.visitWindowFunction(SymbolVisitor.java:50)
	at io.crate.expression.symbol.WindowFunction.accept(WindowFunction.java:70)
	at io.crate.expression.BaseImplementationSymbolVisitor.visitFunction(BaseImplementationSymbolVisitor.java:68)
	at io.crate.expression.BaseImplementationSymbolVisitor.visitFunction(BaseImplementationSymbolVisitor.java:43)
	at io.crate.expression.symbol.Function.accept(Function.java:241)
	at io.crate.expression.InputFactory$Context.add(InputFactory.java:159)
	at io.crate.expression.RowFilter.<init>(RowFilter.java:50)
	at io.crate.expression.RowFilter.create(RowFilter.java:44)
	at io.crate.execution.engine.pipeline.ProjectionToProjectorVisitor.visitFilterProjection(ProjectionToProjectorVisitor.java:509)
	at io.crate.execution.engine.pipeline.ProjectionToProjectorVisitor.visitFilterProjection(ProjectionToProjectorVisitor.java:127)
	at io.crate.execution.dsl.projection.FilterProjection.accept(FilterProjection.java:80)
	at io.crate.execution.dsl.projection.ProjectionVisitor.process(ProjectionVisitor.java:30)
	at io.crate.execution.engine.pipeline.ProjectionToProjectorVisitor.create(ProjectionToProjectorVisitor.java:699)
	at io.crate.execution.engine.pipeline.Projectors.wrap(Projectors.java:83)
	at io.crate.execution.engine.collect.sources.ProjectorSetupCollectSource.lambda$getIterator$0(ProjectorSetupCollectSource.java:50)
	at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:680)
	at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:658)
	at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2164)
	at io.crate.execution.engine.collect.sources.ProjectorSetupCollectSource.getIterator(ProjectorSetupCollectSource.java:50)
	at io.crate.execution.engine.collect.MapSideDataCollectOperation.createIterator(MapSideDataCollectOperation.java:58)
	at io.crate.execution.engine.collect.CollectTask.start(CollectTask.java:162)
	at io.crate.execution.jobs.RootTask.start(RootTask.java:191)
	at io.crate.execution.engine.JobLauncher.setupTasks(JobLauncher.java:241)
	at io.crate.execution.engine.JobLauncher.execute(JobLauncher.java:154)
	at io.crate.planner.operators.LogicalPlanner.executeNodeOpTree(LogicalPlanner.java:531)
	at io.crate.planner.operators.LogicalPlanner.doExecute(LogicalPlanner.java:482)
	at io.crate.planner.operators.LogicalPlanner.execute(LogicalPlanner.java:455)
	at io.crate.planner.operators.LogicalPlan.executeOrFail(LogicalPlan.java:220)
	at io.crate.planner.Plan.execute(Plan.java:74)
	at io.crate.action.sql.Session.singleExec(Session.java:658)
	at io.crate.action.sql.Session.execute(Session.java:466)
	at io.crate.rest.action.SqlHttpHandler.executeSimpleRequest(SqlHttpHandler.java:266)
	at io.crate.rest.action.SqlHttpHandler.handleSQLRequest(SqlHttpHandler.java:206)
	at io.crate.rest.action.SqlHttpHandler.channelRead0(SqlHttpHandler.java:115)
	... 45 more
```
</details>

--- a/docs/appendices/release-notes/unreleased.rst
+++ b/docs/appendices/release-notes/unreleased.rst
@@ -89,6 +89,15 @@ Changes
 Fixes
 =====
 
+- Fixed an issue that resulted in a non-executable plan if a windows function
+  result from a sub-select is used inside a query filter. An example::
+
+      SELECT * FROM (
+        SELECT ROW_NUMBER() OVER(PARTITION by col1) as row_num
+        FROM (VALUES('x')) t1
+      ) t2
+      WHERE row_num = 2;
+
 - Fixed an issue that caused valid values for ``number_of_routing_shards`` in
   ``CREATE TABLE`` statements to be rejected because the validation always used
   a fixed value of ``5`` instead of the actual number of shards declared within

--- a/server/src/main/java/io/crate/planner/operators/TableFunction.java
+++ b/server/src/main/java/io/crate/planner/operators/TableFunction.java
@@ -170,4 +170,13 @@ public final class TableFunction implements LogicalPlan {
             .text(where.queryOrFallback().toString())
             .text("]");
     }
+
+    @Override
+    public String toString() {
+        return "TableFunction{" +
+            "relation=" + relation +
+            ", toCollect=" + toCollect +
+            ", where=" + where +
+            '}';
+    }
 }

--- a/server/src/main/java/io/crate/planner/operators/WindowAgg.java
+++ b/server/src/main/java/io/crate/planner/operators/WindowAgg.java
@@ -21,21 +21,9 @@
 
 package io.crate.planner.operators;
 
-import static io.crate.execution.dsl.phases.ExecutionPhases.executesOnHandler;
-
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.HashSet;
-import java.util.LinkedHashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.UUID;
-import java.util.function.Function;
-
-import javax.annotation.Nullable;
-
 import io.crate.analyze.OrderBy;
 import io.crate.analyze.WindowDefinition;
+import io.crate.common.annotations.VisibleForTesting;
 import io.crate.common.collections.Lists2;
 import io.crate.data.Row;
 import io.crate.execution.dsl.phases.MergePhase;
@@ -55,6 +43,18 @@ import io.crate.planner.distribution.DistributionInfo;
 import io.crate.planner.distribution.DistributionType;
 import io.crate.statistics.TableStats;
 
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.function.Function;
+
+import static io.crate.execution.dsl.phases.ExecutionPhases.executesOnHandler;
+
 public class WindowAgg extends ForwardingLogicalPlan {
 
     final WindowDefinition windowDefinition;
@@ -62,7 +62,8 @@ public class WindowAgg extends ForwardingLogicalPlan {
     private final List<Symbol> standalone;
     private final List<Symbol> outputs;
 
-    static LogicalPlan create(LogicalPlan source, List<WindowFunction> windowFunctions) {
+    @VisibleForTesting
+    public static LogicalPlan create(LogicalPlan source, List<WindowFunction> windowFunctions) {
         if (windowFunctions.isEmpty()) {
             return source;
         }
@@ -126,7 +127,7 @@ public class WindowAgg extends ForwardingLogicalPlan {
         return new WindowAgg(newSource, windowDefinition, List.copyOf(newWindowFunctions), newSource.outputs());
     }
 
-    List<WindowFunction> windowFunctions() {
+    public List<WindowFunction> windowFunctions() {
         return windowFunctions;
     }
 
@@ -239,6 +240,10 @@ public class WindowAgg extends ForwardingLogicalPlan {
 
     @Override
     public String toString() {
-        return "WindowAgg{[" + Lists2.joinOn(", ", windowFunctions, WindowFunction::toString) + "]}";
+        return "WindowAgg{" +
+            "source=" + source + ", " +
+            "windowDefinition=" + windowDefinition + ", " +
+            "windowFunctions=[" + Lists2.joinOn(", ", windowFunctions, WindowFunction::toString) + "]" +
+            "}";
     }
 }

--- a/server/src/main/java/io/crate/planner/optimizer/rule/MoveFilterBeneathWindowAgg.java
+++ b/server/src/main/java/io/crate/planner/optimizer/rule/MoveFilterBeneathWindowAgg.java
@@ -22,9 +22,12 @@
 package io.crate.planner.optimizer.rule;
 
 import io.crate.analyze.WindowDefinition;
+import io.crate.expression.operator.AndOperator;
+import io.crate.expression.symbol.Symbol;
+import io.crate.expression.symbol.SymbolVisitors;
+import io.crate.expression.symbol.WindowFunction;
 import io.crate.metadata.NodeContext;
 import io.crate.metadata.TransactionContext;
-import io.crate.statistics.TableStats;
 import io.crate.planner.operators.Filter;
 import io.crate.planner.operators.LogicalPlan;
 import io.crate.planner.operators.WindowAgg;
@@ -32,6 +35,11 @@ import io.crate.planner.optimizer.Rule;
 import io.crate.planner.optimizer.matcher.Capture;
 import io.crate.planner.optimizer.matcher.Captures;
 import io.crate.planner.optimizer.matcher.Pattern;
+import io.crate.statistics.TableStats;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Predicate;
 
 import static io.crate.planner.operators.LogicalPlanner.extractColumns;
 import static io.crate.planner.optimizer.matcher.Pattern.typeOf;
@@ -55,16 +63,84 @@ public final class MoveFilterBeneathWindowAgg implements Rule<Filter> {
     }
 
     @Override
-    public LogicalPlan apply(Filter plan,
+    public LogicalPlan apply(Filter filter,
                              Captures captures,
                              TableStats tableStats,
                              TransactionContext txnCtx,
                              NodeContext nodeCtx) {
         WindowAgg windowAgg = captures.get(windowAggCapture);
         WindowDefinition windowDefinition = windowAgg.windowDefinition();
-        if (windowDefinition.partitions().containsAll(extractColumns(plan.query()))) {
-            return transpose(plan, windowAgg);
+        List<WindowFunction> windowFunctions = windowAgg.windowFunctions();
+
+        Predicate<Symbol> containsWindowFunction =
+            symbol -> symbol instanceof WindowFunction && windowFunctions.contains(symbol);
+
+        Symbol predicate = filter.query();
+        List<Symbol> filterParts = AndOperator.split(predicate);
+
+        ArrayList<Symbol> remainingFilterSymbols = new ArrayList<>();
+        ArrayList<Symbol> windowPartitionedBasedFilters = new ArrayList<>();
+
+        for (Symbol part : filterParts) {
+            if (SymbolVisitors.any(containsWindowFunction, part) == false
+                && windowDefinition.partitions().containsAll(extractColumns(part))) {
+                windowPartitionedBasedFilters.add(part);
+            } else {
+                remainingFilterSymbols.add(part);
+            }
         }
-        return null;
+        assert remainingFilterSymbols.size() > 0 || windowPartitionedBasedFilters.size() > 0 :
+            "Splitting the filter symbol must result in at least one group";
+
+        /* SELECT ROW_NUMBER() OVER(PARTITION BY id)
+         * WHERE `x = 1`
+         *
+         * `x` is not the window partition column.
+         * We cannot push down the filter as it would change the window aggregation value
+         *
+         */
+        if (windowPartitionedBasedFilters.isEmpty()) {
+            return null;
+        }
+
+        /* SELECT ROW_NUMBER() OVER(PARTITION BY id)
+         * WHERE `id = 1`
+         * remainingFilterSymbols:                  []
+         * windowPartitionedBasedFilters:           [id = 1]
+         *
+         * Filter
+         *  |
+         * WindowsAgg
+         *
+         * transforms to
+         *
+         * WindowAgg
+         *  |
+         * Filter (id = 1)
+         */
+        if (remainingFilterSymbols.isEmpty()) {
+            return transpose(filter, windowAgg);
+        }
+
+        /* WHERE `ROW_NUMBER() OVER(PARTITION BY id) = 2 AND id = 1`
+         * remainingFilterSymbols:                  [ROW_NUMBER() OVER(PARTITION BY id) = 2]
+         * windowPartitionedBasedFilters:           [id = 1]
+         *
+         * Filter
+         *  |
+         * WindowsAgg
+         *
+         * transforms to
+         *
+         * Filter (ROW_NUMBER() OVER(PARTITION BY id) = 2)
+         *  |
+         * WindowAgg
+         *  |
+         * Filter (id = 1)
+         */
+        LogicalPlan newWindowAgg = windowAgg.replaceSources(
+            List.of(new Filter(windowAgg.source(), AndOperator.join(windowPartitionedBasedFilters))));
+
+        return new Filter(newWindowAgg, AndOperator.join(remainingFilterSymbols));
     }
 }

--- a/server/src/test/java/io/crate/planner/optimizer/rule/MoveFilterBeneathWindowAggTest.java
+++ b/server/src/test/java/io/crate/planner/optimizer/rule/MoveFilterBeneathWindowAggTest.java
@@ -1,0 +1,200 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.planner.optimizer.rule;
+
+import io.crate.expression.symbol.Symbol;
+import io.crate.expression.symbol.WindowFunction;
+import io.crate.metadata.CoordinatorTxnCtx;
+import io.crate.planner.operators.Filter;
+import io.crate.planner.operators.LogicalPlan;
+import io.crate.planner.operators.WindowAgg;
+import io.crate.planner.optimizer.matcher.Captures;
+import io.crate.planner.optimizer.matcher.Match;
+import io.crate.statistics.TableStats;
+import io.crate.test.integration.CrateDummyClusterServiceUnitTest;
+import io.crate.testing.SQLExecutor;
+import org.hamcrest.Matchers;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.List;
+
+import static io.crate.planner.operators.LogicalPlannerTest.isPlan;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.nullValue;
+
+public class MoveFilterBeneathWindowAggTest extends CrateDummyClusterServiceUnitTest {
+
+    private SQLExecutor e;
+
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+        e = SQLExecutor.builder(clusterService)
+            .addTable("create table t1 (id int, x int)")
+            .build();
+    }
+
+    @Test
+    public void test_filter_on_containing_windows_function_is_not_moved() {
+        var collect = e.logicalPlan("SELECT id FROM t1");
+
+        WindowFunction windowFunction = (WindowFunction) e.asSymbol("ROW_NUMBER() OVER(PARTITION by id)");
+        WindowAgg windowAgg = (WindowAgg) WindowAgg.create(collect, List.of(windowFunction));
+
+        Symbol query = e.asSymbol("ROW_NUMBER() OVER(PARTITION by id) = 2");
+        Filter filter = new Filter(windowAgg, query);
+
+        var rule = new MoveFilterBeneathWindowAgg();
+        Match<Filter> match = rule.pattern().accept(filter, Captures.empty());
+
+        assertThat(match.isPresent(), is(true));
+        assertThat(match.value(), Matchers.sameInstance(filter));
+
+        LogicalPlan newPlan = rule.apply(
+            match.value(),
+            match.captures(),
+            new TableStats(),
+            CoordinatorTxnCtx.systemTransactionContext(),
+            e.nodeCtx
+        );
+
+        assertThat(newPlan, nullValue());
+    }
+
+    @Test
+    public void test_filter_containing_columns_not_part_of_the_window_partition_cannot_be_moved() {
+        var collect = e.logicalPlan("SELECT id, x FROM t1");
+
+        WindowFunction windowFunction = (WindowFunction) e.asSymbol("ROW_NUMBER() OVER(PARTITION by id)");
+        WindowAgg windowAgg = (WindowAgg) WindowAgg.create(collect, List.of(windowFunction));
+
+        Symbol query = e.asSymbol("x = 1");
+        Filter filter = new Filter(windowAgg, query);
+
+        var rule = new MoveFilterBeneathWindowAgg();
+        Match<Filter> match = rule.pattern().accept(filter, Captures.empty());
+
+        assertThat(match.isPresent(), is(true));
+        assertThat(match.value(), Matchers.sameInstance(filter));
+
+        LogicalPlan newPlan = rule.apply(
+            match.value(),
+            match.captures(),
+            new TableStats(),
+            CoordinatorTxnCtx.systemTransactionContext(),
+            e.nodeCtx
+        );
+
+        assertThat(newPlan, nullValue());
+    }
+
+    @Test
+    public void test_filter_on_windows_function_partition_columns_is_moved() {
+        var collect = e.logicalPlan("SELECT id FROM t1");
+
+        WindowFunction windowFunction = (WindowFunction) e.asSymbol("ROW_NUMBER() OVER(PARTITION BY id)");
+        WindowAgg windowAgg = (WindowAgg) WindowAgg.create(collect, List.of(windowFunction));
+
+        Symbol query = e.asSymbol("id = 10");
+        Filter filter = new Filter(windowAgg, query);
+
+        var rule = new MoveFilterBeneathWindowAgg();
+        Match<Filter> match = rule.pattern().accept(filter, Captures.empty());
+
+        assertThat(match.isPresent(), is(true));
+        assertThat(match.value(), Matchers.sameInstance(filter));
+
+        LogicalPlan newPlan = rule.apply(
+            match.value(),
+            match.captures(),
+            new TableStats(),
+            CoordinatorTxnCtx.systemTransactionContext(),
+            e.nodeCtx
+        );
+        var expectedPlan =
+            "WindowAgg[id, row_number() OVER (PARTITION BY id)]\n" +
+            "  └ Filter[(id = 10)]\n" +
+            "    └ Collect[doc.t1 | [id] | true]";
+
+        assertThat(newPlan, isPlan(expectedPlan));
+    }
+
+    @Test
+    public void test_filter_part_on_windows_function_partition_columns_is_moved() {
+        var collect = e.logicalPlan("SELECT id FROM t1");
+
+        WindowFunction windowFunction = (WindowFunction) e.asSymbol("ROW_NUMBER() OVER(PARTITION BY id)");
+        WindowAgg windowAgg = (WindowAgg) WindowAgg.create(collect, List.of(windowFunction));
+
+        Symbol query = e.asSymbol("ROW_NUMBER() OVER(PARTITION BY id) = 2 AND id = 10 AND x = 1");
+        Filter filter = new Filter(windowAgg, query);
+
+        var rule = new MoveFilterBeneathWindowAgg();
+        Match<Filter> match = rule.pattern().accept(filter, Captures.empty());
+
+        assertThat(match.isPresent(), is(true));
+        assertThat(match.value(), Matchers.sameInstance(filter));
+
+        LogicalPlan newPlan = rule.apply(
+            match.value(),
+            match.captures(),
+            new TableStats(),
+            CoordinatorTxnCtx.systemTransactionContext(),
+            e.nodeCtx
+        );
+        var expectedPlan =
+            "Filter[((row_number() OVER (PARTITION BY id) = 2) AND (x = 1))]\n" +
+            "  └ WindowAgg[id, row_number() OVER (PARTITION BY id)]\n" +
+            "    └ Filter[(id = 10)]\n" +
+            "      └ Collect[doc.t1 | [id] | true]";
+
+        assertThat(newPlan, isPlan(expectedPlan));
+    }
+
+    @Test
+    public void test_filters_combined_by_OR_cannot_be_moved() {
+        var collect = e.logicalPlan("SELECT id FROM t1");
+
+        WindowFunction windowFunction = (WindowFunction) e.asSymbol("ROW_NUMBER() OVER(PARTITION BY id)");
+        WindowAgg windowAgg = (WindowAgg) WindowAgg.create(collect, List.of(windowFunction));
+
+        Symbol query = e.asSymbol("ROW_NUMBER() OVER(PARTITION BY id) = 1 OR id = 10");
+        Filter filter = new Filter(windowAgg, query);
+
+        var rule = new MoveFilterBeneathWindowAgg();
+        Match<Filter> match = rule.pattern().accept(filter, Captures.empty());
+
+        assertThat(match.isPresent(), is(true));
+        assertThat(match.value(), Matchers.sameInstance(filter));
+
+        LogicalPlan newPlan = rule.apply(
+            match.value(),
+            match.captures(),
+            new TableStats(),
+            CoordinatorTxnCtx.systemTransactionContext(),
+            e.nodeCtx
+        );
+        assertThat(newPlan, nullValue());
+    }
+}

